Add support for WCH CH56x series SoC

xingrz opened the pull request:
**QingKe** is a series of RISC-V cores developed by WCH, consisting with V2A, V3A and V4B/C/F series of MCUs.

CH569/CH565 are 2 of the members of the QingKe V3A series. This PR brings initial support for these two SoCs to Zephyr.

In summary, this PR includes:

* An interrupt controller for all QingKe MCUs.
* A systick driver for the QingKe V3 series.
* Clock controller and GPIO, UART driver specifically for the CH56x MCUs.
* Configurations for the official `CH565W-EVT` and `CH569W-EVT` evaluation boards

Support for other series members will be submit in future PRs.

References:

* https://www.wch-ic.com/products/CH569.html

## TODOs

- [x] #65772
- [x] Get ready for HWMv2
- [ ] Support for custom interrupt controller, either #63214 or #66505

xingrz committed:
dts/soc: wch: Add WCH QingKe V3 series

QingKe V3 is an RV32IMAC core developed by WCH

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts/soc: wch: ch56x: Add WCH CH569/5 SoC

Based on QingKe V3A core, CH569/5 are members of the CH56x series.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: vendor-prefixes: Add WCH

* Website: https://wch-ic.com/

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: timer: Add systick driver for WCH QingKe V3 CPUs

This driver is shared across QingKe V3A series, including CH32V103,
CH569/5 and CH573/1.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: intc: Implement interrupt controller for WCH QingKe CPUs

PFIC (Programmable Fast Interrupt Controller) is the interrupt controller
implmentation in WCH QingKe series RISC-V MCUs.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

# Conflicts:
#	drivers/interrupt_controller/CMakeLists.txt
#	drivers/interrupt_controller/Kconfig

xingrz committed:
boards: wch: Bring up WCH CH56(5/9)W-EVT

CH565W-EVT and CH569W-EVT are boards by WCH for the evaluation of
CH565W and CH569W chips.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: gpio: Add GPIO driver for WCH CH56x

This commit implements a GPIO driver for CH56x series.

The 2 instances of GPIO share a same IRQ line on this chip. So the config
`SHARED_IRQ` is selected as well.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: pinctrl: Implement pin controller for WCH CH5xx series

This commit introduces a pinctrl driver for the CH5xx series.

The limitations of pin multiplexing of WCH CH5xx series are described
in the binding configuration.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
scripts: runners: Add wrapper for WCHISPTool

`WCHISPTool_CMD` is the commandline version of WCHISPTool, which is a USB
flashing tool for WCH series MCUs.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
MAINTAINERS: Add WCH platform and assign myself as maintainer

* Add `WCH Platforms` section
* Add `xingrz` to maintainers of `WCH Platform`

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: serial: Implement UART driver for WCH CH5xx series

This commit introduces UART driver for CH5xx series, supporting both
polling and interrupt-driven mode.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

fkokosinski commented on drivers/interrupt_controller/intc_qingke_pfic.c:
```diff
@@ -0,0 +1,53 @@
+	uint32_t regval;
+
+	regval = sys_read32(PFIC_ITHRESDR);
+	sys_write32(1 << 4, PFIC_ITHRESDR);
```
```suggestion
	sys_write32(BIT(4), PFIC_ITHRESDR);
```

fkokosinski commented on dts/bindings/clock/wch,ch56x-hclk.yaml:
```diff
@@ -0,0 +1,33 @@
+  clock-frequency:
+    type: int
+    description: |
+      System clock frequency in Hz. Supported values are:
```
Can we use MHz instead? The granularity seems to be 1 MHz anyway.

xingrz replied:
I'm not sure. I noticed that another driver (nuvoton,npcx-pcc.yaml) also uses an enum in Hz. I think that most people would naturally assume that a field named `clock-frequency` is expressed in Hz, unless it is explicitly named something like `clock-frequency-mhz`.

fkokosinski commented on dts/bindings/clock/wch,ch56x-hclk.yaml:
```diff
@@ -0,0 +1,33 @@
+        - 80000000
+        - 96000000
+        - 120000000
+    default: 15000000
```
Note: default values need to be justified as per Zephyr's requirements.

fkokosinski commented on include/zephyr/dt-bindings/gpio/ch56x-gpio.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef ZEPHYR_DT_CH56X_GPIO_H_
+#define ZEPHYR_DT_CH56X_GPIO_H_
+
+#define CH56X_GPIO_DRIVE_STRENGTH_16MA (1 << 8)
```
```suggestion
#define CH56X_GPIO_DRIVE_STRENGTH_16MA BIT(8)
```

fkokosinski commented on include/zephyr/dt-bindings/gpio/ch56x-gpio.h:
```diff
@@ -0,0 +1,12 @@
+#define ZEPHYR_DT_CH56X_GPIO_H_
+
+#define CH56X_GPIO_DRIVE_STRENGTH_16MA (1 << 8)
+#define CH56X_GPIO_SCHMITT_TRIGGER     (1 << 9)
```
```suggestion
#define CH56X_GPIO_SCHMITT_TRIGGER     BIT(9)
```

fkokosinski commented on boards/riscv/ch569w_evt/ch569w_evt_defconfig:
```diff
@@ -0,0 +1,7 @@
+CONFIG_SOC_SERIES_CH56X=y
+CONFIG_SOC_CH569=y
+CONFIG_BOARD_CH569W_EVT=y
+CONFIG_BUILD_OUTPUT_HEX=y
```
Let's add a comment saying that this option is required for the custom runner.

fkokosinski replied:
Here and in the other board as well

fkokosinski reviewed (commented):
LGTM in general, just some nitpicks

soburi commented on soc/riscv/wch_qingke_v3/common/soc_irq.S:
```diff
@@ -0,0 +1,61 @@
+/*
```
This file seems a clone of `soc/riscv/riscv-privileged/common/soc_irq.S`.
Can this directory be put into `soc/riscv/riscv-privileged/`?

xingrz replied:
I'm afraid no because it uses a different implementation of interrupt controller (`arch_irq_enable`, `arch_irq_disable`, and so).

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/blob/e7340282a7950d4cb071e77faafc249aaa569a8b/soc/riscv/riscv-privileged/common/soc_common_irq.c#L17C5-L38C1

Please add conditions here and integrate directories.

xingrz replied:
Would you give me an advice on the naming of the Kconfig option? It also need a custom implementation of `arch_cpu_idle` (removing the line https://github.com/zephyrproject-rtos/zephyr/blob/bc43e890fd7a6035438a47b8853b6fe652472da2/soc/riscv/riscv-privileged/common/idle.c#L21).

xingrz replied:
Solved by implementing the intc driver as a CLIC. Also introduced a `CONFIG_RISCV_WFI_ON_CPU_IDLE`, explained in message of commit 4a5c732245.

xingrz commented:
Temporarily convert to draft as I have a new idea on the clock_control driver for better flexibility for other WCH SOCs. I'll update the PR as soon as ready.

xingrz committed:
dts/soc: wch: ch56x: Add syscon node for WCH CH56x

This commit describes the system control block of the CH56x in DTS. This
block will be utilized in future drivers for functions such as pin
remapping, clock control, and interrupt management.

Additionally, macros has been added to facilitate easy access to the block.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: clock: Implement system clock mux controller for CH56x

The CH56x can be configured to run on a 30MHz external oscillator or a
480MHz PLL generated by the USB PHY.

This commit introduces a clock controller that can automatically select
one of the two available clock sources and determine the appropriate
divisor based on the given frequency.

The `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` value is also updated with the
selected frequency.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

# Conflicts:
#	drivers/clock_control/Kconfig
#	soc/riscv/riscv-privileged/wch_ch32v/Kconfig.defconfig.ch56x

xingrz commented:
Updated. I think it's ready for review now.

soburi commented on soc/riscv/riscv-privileged/qingke/Kconfig.soc:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_QINGKE
```
Is this processor not support XIP?

xingrz replied:
`XIP` is implied in `config RISCV` in arch/Kconfig. Do I need to explicitly select it in SoC's Kconfig?

soburi replied:
OK. No problem.

soburi commented on drivers/clock_control/clock_control_ch56x_clkmux.c:
```diff
@@ -0,0 +1,136 @@
+static int ch56x_clkmux_get_rate(const struct device *dev, clock_control_subsys_t sys,
+				 uint32_t *rate)
+{
+	const struct ch56x_clkmux_config *cfg = DEV_CFG(dev);
```
It can write as 

```
const struct ch56x_clkmux_config *cfg = dev->config;
```

Rewrite all and remove macro.

soburi commented on include/zephyr/drivers/syscon/qingke.h:
```diff
@@ -0,0 +1,43 @@
+/*
```
Sorry. I'd misunderstood.
I have got the SoC has the 'System Control register' at 0x40001000.
My pointing out is wrong.

The header place to  'include/zephyr/drivers/syscon/ is correct.

I apologize for the inconvenience, but could you please restore it?

So that, Definitions related to system control registers that are scattered all over the place should be grouped together in the header. I will point this out separately.

xingrz replied:
Updated.

xingrz commented on drivers/clock_control/clock_control_ch56x_clkmux.c:
```diff
@@ -0,0 +1,134 @@
+#include <zephyr/drivers/syscon.h>
+#include <soc.h>
+
+#define SYSCON_CLK_PLL_DIV(base)  (base + 0x00)
+#define SYSCON_CLK_CFG_CTRL(base) (base + 0x02)
+
+/* SYSCON_CLK_PLL_DIV Register */
+#define SYSCON_CLK_PLL_DIV_KEY (0x40)
+
+/* SYSCON_CLK_CFG_CTRL Register */
+#define SYSCON_CLK_CFG_CTRL_KEY (0x80)
+#define SYSCON_CLK_SEL_PLL      BIT(1)
+#define SYSCON_CLK_SEL_OSC      (0)
```
Although this system control block is shared, it's not entirely identical across all members of the series. Some registers in this block are consistent across series members but in different offsets (for example, in the [upcoming pinctrl driver](https://github.com/xingrz/zephyr/commit/788fcde1307d86883eb8845ddf8b0b142c5bd08b) that I will submit in a later PR, the `SAFE_ACCESS_SIG` used in `syscon_unlock()`/`syscon_relock()`), while others are completely specific to individual SoCs (such as GPIO interrupts and clock mux in this PR).

My approach is to define these specific registers as children nodes of the `syscon` block, using the nodes' address cell to denote their offset relative to the `syscon` block (like [this](https://github.com/xingrz/zephyr/blob/ea7e5c9dfd82f9abb4d07d44beb7ae6ddc872a49/dts/riscv/wch/ch56x/ch56x-common.dtsi#L37) and [this](https://github.com/xingrz/zephyr/blob/ea7e5c9dfd82f9abb4d07d44beb7ae6ddc872a49/dts/riscv/wch/ch58x/ch58x-common.dtsi#L52)). This approach allows common drivers like pinctrl and per-device clock_control to be shared across all QingKe series SoCs, while SoC-specific usages remain isolated within SoC-specific drivers like GPIO and clock mux.

xingrz commented:
Rebased to solve the merge conflicts

carlocaione commented on soc/riscv/riscv-privileged/qingke/linker.ld:
this should be gone after https://github.com/zephyrproject-rtos/zephyr/pull/64544

xingrz replied:
Thanks. Rebased and updated.

xingrz commented on soc/riscv/riscv-privileged/qingke/soc.h:
It's needed by `include/zephyr/arch/riscv/irq.h`

soburi commented on drivers/interrupt_controller/intc_qingke_pfic.c:
```diff
@@ -0,0 +1,63 @@
+#define PFIC_IRQN_GROUP(irqn) (irqn >> 5)
+#define PFIC_IRQN_SHIFT(irqn) (irqn & BIT_MASK(5))
+
+void riscv_clic_irq_enable(unsigned int irq)
```
QingKe PFIC is not "CLIC".

Besides with this change, 
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/riscv/riscv-privileged/common/soc_common_irq.c#L18-L39

should be better to modify for generalizing.
At this point, I want to hear @carlocaione opinions.

amr-sc replied:
Hi @soburi , @xingrz ,

We also faced this recently: https://github.com/zephyrproject-rtos/zephyr/pull/62606
Looks like we indeed need to come-up with some common API to be used with our custom interrupt controllers.
https://github.com/zephyrproject-rtos/zephyr/pull/63214 is targeting this API

soburi commented on soc/riscv/riscv-privileged/qingke/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
```
The QingKe is SoC __core__ name.
Isn't the SoC name ch5xx?

xingrz replied:
I believe "QingKe" is used both as the name for the their RISC-V core (usually referred to as "QingKe Vxx cores") and the RISC-V product line (usually referred to as "QingKe MCU series").

In summary, WCH has three product lines: "CH32Fxxx" for Cortex-M cores with STM32-compatible peripherals, designed as a pin-to-pin replacement for STM32; "CH32Vxxx" for "QingKe RISC-V core" with STM32-compatible peripherals; and "CH5xx" for "QingKe RISC-V core" with self-developed peripherals.

soburi replied:
If so as, it should be better to use "wch_ch5xx" here.
The directories is classified by product name.
And append manufacturer name as prefix.

xingrz replied:
Thank you for your suggestion. So, if someone wants to contribute support for the CH32V series in the future, would it be appropriate to create a separate `wch_ch32v` directory? In the directory, there's a `Kconfig.series.qingke_v3`. Originally, my idea was that we could add `Kconfig.series.qingke_v4`, `Kconfig.series.qingke_v2`, and so on in the future. These different `series` use various combinations of `RISCV_ISA_EXT_*` flags, and these configurations would also be needed for the CH32V.

xingrz replied:
After doing some research (and asking a friend working at WCH), I think I should pick `wch_ch32` as the soc name. The CH5xx (as well as the later CH6xx) are in fact a special enhanced model of their CH32V common MCU series.

soburi replied:
Sorry for late response.

> After doing some research (and asking a friend working at WCH), I think I should pick wch_ch32 as the soc name. The CH5xx (as well as the later CH6xx) are in fact a special enhanced model of their CH32V common MCU series.

It sounds good.
However, is the `wch_ch32` is not `wch_ch32v`?
WCH has `ch32` series as ARM product line. https://www.wch-ic.com/products/CH32F103.html
I think it needs to append `v` postfix.
WCH has `ch32` series as ARM product line. https://www.wch-ic.com/products/CH32F103.html

xingrz replied:
You're right. Updated.

xingrz commented:
I'd like to further discuss the usage of syscon with all of you @soburi @carlocaione 

As I mentioned in my previous response during the review, this SoC features a system control block where various parameters for different peripherals are configured.

However, this region varies slightly among different series members. Some registers have the same usage but are located at different offsets, while others are entirely different. Even among related functions (such as clock source selection and division), there may be other functional registers interspersed among them (like GPIO interrupts).

Therefore, I intend to represent different registers as child nodes under the syscon node. The addresses of these child nodes, when added to the address of the syscon node, correspond to the memory addresses of the registers. Some of these child nodes will have `compatible` properties to attach the corresponding drivers, while others are solely for address labeling. Is this an abuse of the syscon driver?

Furthermore, I think `"syscon"` it not perfectly suitable for describing this system control block because although most registers are accessed using 8 bits, there are also some that require 16-bit or 32-bit access. However, syscon can only be restricted to one type using the `reg-io-width` property. Therefore, I'm considering whether it should be represented as a `"simple-bus"` node, using regular `sys_readN()` and `sys_writeN()` accesses.

I look forward to hearing your suggestions.

xingrz commented:
Just pushed an update for the syscon things.

Upon careful consideration, I've realized that it's more flexible to define the offsets of registers in a per-soc header rather than relying on DTS addresses extensively. The driver implementation involved remains shared.

I've also decided to move away from using the syscon driver in favor of using `sys_writeN()` and `sys_readN()` functions, taking into account the varying bit widths of register offsets that I mentioned previously.

Looks better.

xingrz committed:
[DNM] arch: riscv: Add support for custom interrupt controller

NOTE: Duplicated with an existing PR. Leaving this commit here just as
a placeholder. Will be rebased soon after that existing PR get merged.

==========================================================================

In certain cases, platforms may have their own implemented interrupt
controllers (e.g., the PFIC on WCH MCUs) that require the ability to
implement the irq_enable/disable functions at the SoC level.

Similar support already exists for ARM and ARM64. This commit adds support
for RISC-V as well.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

# Conflicts:
#	include/zephyr/arch/riscv/irq.h

xingrz commented:
Rebased to solve the merge conflicts

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -28,6 +28,16 @@ config RISCV_GP
+config RISCV_CUSTOM_INTERRUPT_CONTROLLER
+	bool
+	help
+	  This option indicates that the ARM CPU is connected to a custom
```
```suggestion
	  This option indicates that the CPU is connected to a custom
```

Not an ARM CPU ;)

xingrz replied:
Thanks. I expect this commit to be removed soon after #63214 is merged. I added it as a quick fix to make my branch buildable since the solution in that PR is still being discussed.

xingrz commented:
I'm waiting for #66505 to get merged.

soburi commented on soc/riscv/riscv-privileged/wch_ch32v/Kconfig.series.qingke_v3:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_CH32V_QINGKE_V3A
```
This config should be better to merge into `SOC_SERIES_CH32V`.
There are no settings where SOC_SERIES is divided and configured at this time.
And the core type can identify from the devicetree definitions.

xingrz replied:
This config is used as a "shortcut" to select supported ISA extentions. In the further PR I'll submit the support for QingKe V4A and QingKe V4F series, which have different ISA extentions.

soburi replied:
No RISCV SoC makes such a "shortcut".
I think it's better to follow custom.

nandojve commented:
Hi @xingrz ,

I'm not sure if you are aware of #50305 but it is place now.

xingrz commented:
> Hi @xingrz ,
> 
> I'm not sure if you are aware of #50305 but it is place now.

Thank you for reminding me.

I think I should convert this PR to draft again until I finished refactored for the new SoC scheme and the interrupt controller API is ready.

bjarki-trackunit commented:
@xingrz

> I think I should convert this PR to draft again until I finished refactored for the new SoC scheme and the interrupt controller API is ready.

 Working on it :D

nordicjm reviewed (changes requested):
Needs to be migrated to hwmv2

xingrz commented:
> Needs to be migrated to hwmv2

Yes I'm planning to do that. I'm still waiting for the custom interrupt controller before this PR can go ahead.

Colt-Ma commented:
@xingrz @bjarki-trackunit
Is this still progressing? Qingke series microcontrollers will definitely be used in large quantities because of their low price, so I think it is necessary to adapt this series of microcontrollers to zephyr.

Colt-Ma commented:
@xingrz @bjarki-trackunit
Is this still progressing? Qingke series microcontrollers will definitely be used in large quantities because of their low price, so I think it is necessary to adapt this series of microcontrollers to zephyr.

xingrz commented:
> @xingrz @bjarki-trackunit Is this still progressing? Qingke series microcontrollers will definitely be used in large quantities because of their low price, so I think it is necessary to adapt this series of microcontrollers to zephyr.

Yes, but I'm still waiting for the support for custom interrupt controller  https://github.com/zephyrproject-rtos/zephyr/pull/63214#issuecomment-1842561320

Colt-Ma commented:
Are you Chinese? We can add WeChat, because I'm thinking about adapting to the SoC of the CH59X series

--------------原始邮件--------------
发件人："XiNGRZ ***@***.***&gt;;
发送时间：2024年5月19日(星期天) 下午5:24
收件人："zephyrproject-rtos/zephyr" ***@***.***&gt;;
抄送："Colt-Ma ***@***.***&gt;;"Comment ***@***.***&gt;;
主题：Re: [zephyrproject-rtos/zephyr] Add support for WCH CH56x series SoC (PR #64472)
-----------------------------------

 

  
@xingrz @bjarki-trackunit Is this still progressing? Qingke series microcontrollers will definitely be used in large quantities because of their low price, so I think it is necessary to adapt this series of microcontrollers to zephyr.
  
Yes, but I'm still waiting for the support for custom interrupt controller  #63214 (comment)
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***&gt;

xingrz commented:
> Are you Chinese? We can add WeChat, because I'm thinking about adapting to the SoC of the CH59X series

You can find WIP support for CH56x~CH59x at my repository: https://github.com/xingrz/zephyr/commits/wch/wch-ch59x

Colt-Ma commented:
> > Are you Chinese? We can add WeChat, because I'm thinking about adapting to the SoC of the CH59X series
> 
> You can find WIP support for CH56x~CH59x at my repository: https://github.com/xingrz/zephyr/commits/wch/wch-ch59x

Thank you so much, are you an employee of WCH?

Colt-Ma commented:
@xingrz
I will output the uart0 of ch592f as the console, with a baud rate of 115200. However, the actual baud rate seen on the oscilloscope is 114khz. I am not sure where the problem is and how to modify it. I look forward to your prompt. Thank you!

xingrz commented:
> @xingrz I will output the uart0 of ch592f as the console, with a baud rate of 115200. However, the actual baud rate seen on the oscilloscope is 114khz. I am not sure where the problem is and how to modify it. I look forward to your prompt. Thank you!

Maybe you need to revert the latest `[WIP] TICKLESS]` commit from that branch.

We're off-topic. You can contact me from the official [Discord](https://discord.com/invite/Ck7jw53nU2).

Colt-Ma commented:
> > @xingrz I will output the uart0 of ch592f as the console, with a baud rate of 115200. However, the actual baud rate seen on the oscilloscope is 114khz. I am not sure where the problem is and how to modify it. I look forward to your prompt. Thank you!
> 
> Maybe you need to revert the latest `[WIP] TICKLESS]` commit from that branch.
> 
> We're off-topic. You can contact me from the official [Discord](https://discord.com/invite/Ck7jw53nU2).

I'm probably out of my mind. The baud rate of 115200 is correct when I saw it on an oscilloscope at 114khz. Your code is fine, haha, sorry.

Colt-Ma commented:
@xingrz
I get an error when compiling to adapt the SPI driver to CH592, as follows:

```
# SPDX-License-Identifier: Apache-2.0

description: WCH SPI controller

compatible: "wch,ch5xx-spi"

include: [spi-controller.yaml, pinctrl-device.yaml]

properties:
  reg:
    required: true

  interrupts:
    required: false

  clocks:
    required: true

  pinctrl-0:
    required: true

  pinctrl-names:
    required: true
```
```
spi0: spi@40004000 {
	compatible = "wch,ch5xx-spi";
	reg = < 0x40004000 0x400 >;
	interrupts = < 0x13 0x5 >;
	clocks = < &hclk 0x1 0x0 >;
	status = "okay";
	#address-cells = < 0x1 >;
	#size-cells = < 0x0 >;
	pinctrl-0 = < &spi0_default >;
	pinctrl-names = "default";
	cs-gpios = < &gpioa 0xc 0x1 >;
};
```
```
#define CREATE_CH5XX_SPI_INSTANCE(num)						\
	static void spi_##num##_irq_init(void)					\
	{														\
		IRQ_CONNECT(DT_INST_IRQN(num),						\
			    DT_INST_IRQ(num, priority),					\
			    spi_isr, DEVICE_DT_INST_GET(num), 0);		\
		irq_enable(DT_INST_IRQN(num));						\
	}														\
	static struct spi_ch5xx_data spi_##num##_data = {		\
		SPI_CONTEXT_INIT_LOCK(spi_##num##_data, ctx),		\
		SPI_CONTEXT_INIT_SYNC(spi_##num##_data, ctx),		\
	};														\
	static const struct spi_ch5xx_config spi_##num##_config = {	\
		.irq_config = spi_##num##_irq_init,					\
		.base = DT_INST_REG_ADDR(num),						\
		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(num),		\
		.ccfg = CH5XX_CLOCK_DT_INST_DEV_CONFIG_GET(num),	\
	};														\
	PM_DEVICE_DEFINE(spi_##num, spi_ch5xx_device_ctrl);		\
	DEVICE_DT_INST_DEFINE(num,								\
			      spi_ch5xx_init,							\
			      PM_DEVICE_GET(spi_##num),					\
			      &spi_##num##_data,						\
			      &spi_##num##_config,						\
			      POST_KERNEL,								\
			      CONFIG_SPI_INIT_PRIORITY,					\
			      &ch5xx_spi_api);

DT_INST_FOREACH_STATUS_OKAY(CREATE_CH5XX_SPI_INSTANCE)

```

goutnet commented:
I am very interested to follow this PR, how can I help @xingrz

soburi commented:
@xingrz 

https://github.com/zephyrproject-rtos/zephyr/pull/66505 seems like a long-term task to finish.
I propose this PR progressing with the current ISR implementation.
What do you think?

cub-uanic commented:
Any plans to merge it?
Any plans to support CH58x and CH59x?
Thanks!
